[WIP] 1.22 feature blog for api server tracing (#28991)
kubernetes/website: content/en/blog/_posts/2021-08-06-api-server-tracing.md

---
layout: blog
title: 'API Server Tracing'
date: 2021-08-06
slug: api-server-tracing
---

**Authors:** David Ashpole (Google)

> In distributed systems, it can be hard to figure out where problems are. You grep through one components' logs just to discover that the source of your problem is in another component.  You search there only to discover that you need to enable debug logs to figure out what really went wrong... And it goes on. The more complex the path your request takes, the harder it is to answer questions about where it went.  I've personally spent many hours doing this dance with a variety of kubernetes components.

Reviewer: ```suggestion
In distributed systems, it can be hard to figure out where problems are. You grep through one component's logs just to discover that the source of your problem is in another component.  You search there only to discover that you need to enable debug logs to figure out what really went wrong... And it goes on. The more complex the path your request takes, the harder it is to answer questions about where it went.  I've personally spent many hours doing this dance with a variety of Kubernetes components.
```

## What is Tracing?

> Tracing is supposed to help with these cases.  Whereas logs are grouped together for a given component, traces are grouped together by the request that caused them.  Logging limits the amount of data ingested by using log levels, which means details are often missing.  Tracing uses sampling to collect fewer, but generally more detailed traces. My favorite aspect, though, are how useful the visualisations of traces are.  As we will see, even if you don't understand the inner workings of the API Server, or don't have a clue what an Etcd "Transaction" is, I'd wager you can interpret the trace below (Spoiler!), and would have a pretty good guess for where the problem lies.

Reviewer: I would change "My favorite aspect, though, are" to "My favorite aspect, though, is". I know that in the UK people often refer to things in plural where we would refer in the singular. Changing "aspect" to "aspects" could work as well.

Reviewer 2: Updated

## Why OpenTelemetry?

> The some of the Kubernetes community's core values are portability, openness, and vendor neutrality. It's important that Kubernetes works well for everyone, regardless of who manages your infrastructure, or which vendors you choose to integrate with -- that goes doubly for telemetry solutions.  OpenTelemetry, being a CNCF project, shares these core values, and is creating exactly what we need in Kubernetes: A set of open standards for Tracing client library APIs and a standard trace format. By using OpenTelemetry, we can ensure users have the freedom to choose their backend, and ensure vendors have a level playing-field. The timing is right, as well: the OpenTelemetry golang API and SDK are very close their 1.0 release.

Reviewer: For "The some of" do you mean "The sum of" or "Some of"? Also, I think you can drop the hyphen (-) in "playing-field".

Reviewer 2: Fixed.

---
layout: blog
title: 'Alpha in Kubernetes v1.22: API Server Tracing'

> date: 2021-08-23

Reviewer 3: ```suggestion
date: 2021-09-03
```
Proposing Fri, Sept 3 as the new publish date

> 

Reviewer 4: nit: I'd omit this initial white space. I don't think it bothers Hugo though.

Reviewer 2: done

---
layout: blog

> title: 'API Server Tracing'

Reviewer 4: ```suggestion
title: 'Alpha in v1.22: API Server Tracing'
```
?

Reviewer 2: done

date: 2021-08-23
slug: api-server-tracing
---

**Authors:** David Ashpole (Google)

> In distributed systems, it can be hard to figure out where problems are. You grep through one component's logs just to discover that the source of your problem is in another component.  You search there only to discover that you need to enable debug logs to figure out what really went wrong... And it goes on. The more complex the path your request takes, the harder it is to answer questions about where it went.  I've personally spent many hours doing this dance with a variety of Kubernetes components. Distributed tracing is a tool which is designed to help in these situations, and the Kubernetes API Server is, perhaps, the most important Kubernetes component to be able to debug. At Kubernetes' Sig-Instrumentation, our mission is to make it easier to understand what's going on in your cluster, and we are happy to announce that distributed tracing in the Kubernetes API Server reached alpha in 1.22.

Reviewer 4: nit: “SIG Instrumentation” (with a space)

Reviewer 2: done

## What is Tracing?

> Distributed tracing links together a bunch of super-detailed information from multiple different sources, and structures that telemetry into a single tree for that request.  Unlike logging, which limits the quantity of data ingested by using log levels, tracing collects all of the details and uses sampling to collect only a small percentage of requests.  This means that once you have a trace which demonstrates an issue, you should have all the information you need to root-cause the problem--no grepping for object UID required! My favorite aspect, though, is how useful the visualizations of traces are.  Even if you don't understand the inner workings of the API Server, or don't have a clue what an Etcd "Transaction" is, I'd wager you (yes, you!) could tell me roughly what the order of events was, and which components were involved in the request.  If some step takes a long time, it is easy to tell where the problem is.

Reviewer 4: nit: etcd (all lowercase)

Reviewer 2: done

## Why OpenTelemetry?

It's important that Kubernetes works well for everyone, regardless of who manages your infrastructure, or which vendors you choose to integrate with.  That is particularly true for Kubernetes' integrations with telemetry solutions.  OpenTelemetry, being a CNCF project, shares these core values, and is creating exactly what we need in Kubernetes: A set of open standards for Tracing client library APIs and a standard trace format. By using OpenTelemetry, we can ensure users have the freedom to choose their backend, and ensure vendors have a level playing field. The timing couldn't be better: the OpenTelemetry golang API and SDK are very close to their 1.0 release, and will soon offer backwards-compatibility for these open standards.

## Why instrument the API Server?

The Kubernetes API Server is a great candidate for tracing for a few reasons:

* It follows the standard "RPC" model (serve a request by making requests to downstream components), which makes it easy to instrument.
* Users are latency-sensitive: If a request takes more than 10 seconds to complete, many clients will time-out.

> * It has a complex service topology: A single request could require consulting a dozen webhooks, or involve multiple requests to etcd.

Reviewer 5: Is it worth showing https://d33wubrfki0l68.cloudfront.net/af21ecd38ec67b3d81c1b762221b4ac777fcf02d/7c60e/images/blog/2019-03-21-a-guide-to-kubernetes-admission-controllers/admission-controller-phases.png or something similar here, to show the topology explicitly?

Reviewer 2: Where did that come from?  Is that something I can use here?

Reviewer 4: I think it's from https://kubernetes.io/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/#what-are-kubernetes-admission-controllers

If using that, would be nice to switch to a SVG version and / or use Kubernetes colors for the diagram.

Reviewer 5: That image is from https://kubernetes.io/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/.

Reviewer 2: Thanks.  Any recommendations on how I can make it use kubernetes colors (what are the kubernetes colors)?  Should I re-make the diagram, or is there an easy-ish way to edit the image?

Reviewer 4: It's fine not to add a diagram - I'd rather get this published.

## Trying out APIServer Tracing with a webhook

### Enabling API Server Tracing

1. Enable the APIServerTracing [feature-gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/).

2. Set our configuration for tracing by pointing the `--tracing-config-file` flag on the kube-apiserver at our config file, which contains:

```yaml
apiVersion: apiserver.config.k8s.io/v1alpha1
kind: TracingConfiguration
# 1% sampling rate

> samplingRatePerMillion: 10000

Reviewer 4: Isn't that 10%?

Reviewer 2: 10^4/10^6 = 1/100

```

### Enabling Etcd Tracing

Add `--experimental-enable-distributed-tracing`,  `--experimental-distributed-tracing-address=0.0.0.0:4317`, `--experimental-distributed-tracing-service-name=etcd` flags to etcd to enable tracing.  Note that this traces every request, so it will probably generate a lot of traces if you enable it.

### Example Trace: List Nodes

> I could've used any trace backend, but decided to use Jaeger, since it is one of the most popular open-source tracing projects.  I deployed the Jaeger All-in-one in my cluster, and an deployed OpenTelemetry collector on my control-plane node ([example](https://github.com/dashpole/dashpole_demos/tree/master/otel/controlplane)), and captured traces like this one:

Reviewer 5: ```suggestion
I could've used any trace backend, but decided to use Jaeger, since it is one of the most popular open-source tracing projects.  I deployed [the Jaeger All-in-one container](https://hub.docker.com/r/jaegertracing/all-in-one) in my cluster, deployed [the OpenTelemetry collector](https://github.com/open-telemetry/opentelemetry-collector) on my control-plane node ([example](https://github.com/dashpole/dashpole_demos/tree/master/otel/controlplane)), and captured traces like this one:
```

> ![APIServer_Etcd](https://user-images.githubusercontent.com/3262098/128613151-91cb925c-4886-4f05-a12a-771c6cbe9807.png)

Reviewer 4: _this_ image should live in the same Git repository as the article.

We can merge this and then track a separate issue to migrate it over. Also, the alt attribute should be a readable phrase (spaces not underscores).

Reviewer 2: I don't see alt text for this one.  Do I need to add it?

Reviewer 4: This uses Markdown and is currently defining the alt attribute as `APIServer_Etcd`. I'd write “Jaeger screenshot showing API server and etcd trace”.

Reviewer 4: (these are nits - we could merge as-is if technical review is in place, and then open a follow up PR to fix things like this)

Reviewer 2: updated the text to what you suggested.

The teal lines are from the API Server, and includes it serving a request to `/api/v1/nodes`, and issuing a grpc `Range` RPC to ETCD.  The yellow-ish line is from ETCD handling the `Range` RPC.

### Example Trace: Create Pod with Mutating Webhook

I instrumented the [example webhook](https://github.com/kubernetes-sigs/controller-runtime/tree/master/examples/builtins) with OpenTelemetry (I had to [patch](https://github.com/dashpole/controller-runtime/commit/85fdda7ba03dd2c22ef62c1a3dbdf5aa651f90da) controller-runtime, but it makes a neat demo), and routed traces to Jaeger as well.  I collected traces like this one:

> <img width="1440" alt="APIServer_webhook_etcd" src="https://user-images.githubusercontent.com/3262098/128613167-e7a14cdf-5635-422f-9fd8-f32744ce639d.png">

Reviewer 4: _this_ image should live in the Git repository

We can merge this and then track a separate issue to migrate it over. Also, the alt attribute should be a readable phrase (spaces not underscores).

Reviewer 2: What do you mean by the "alt" attribute?

Reviewer 2: Updated alt text
